Add sql-format extension

tonngw opened the pull request:
## Description

<!-- A summary of your change. If you add a new extension or command, explain what it does. -->
This extension is used to format SQL clause in Raycast. Support format a SQL cluase string stored in clipboard and selected in the foremost editor, then copies or pasted it back.

This tool is simple and easy to use. I believe it will become a must-have tool for many developers every day.

## Screencast

<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

![Raycast 2024-08-24 at 22 23 04](https://github.com/user-attachments/assets/658979ff-7f2e-4cbc-93bb-f80ecd86889c)


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

jfkisafk commented on extensions/sql-format/assets/extension-icon.png:
We can go with something like this to be more specific:

![extension_icon (4)](https://github.com/user-attachments/assets/98766ca9-1607-4dd2-99c0-826112c8c5be)

tonngw replied:
yeah, i look for many times about sql icon, but no good-looking icons found, so check it a generic icon haha.

jfkisafk replied:
Yeah the above icon is the best I could come up with too. If you like it just save it to assets dir. It is already 512px. Otherwise, you can resolve the comment.

jfkisafk commented on extensions/sql-format/src/index.tsx:
```diff
@@ -0,0 +1,179 @@
+  }
+}
+
+// user preferences settings
+interface Preferences {
+  tabWidth: tabWidthType;
+  keywordCase: keywordCaseType;
+  autopaste: boolean;
+}
+
+type tabWidthType = '2' | '4' | 'tab';
+type keywordCaseType = 'preserve' | 'upper' | 'lower';
```
(non-blocking) These are not necessary, as _Preferences_ is present in _raycast-env.d.ts_. See https://github.com/raycast/extensions/pull/14155 for more details.

```suggestion
```

tonngw replied:
I try it works, thank you tell for me the tricks. I really appreciate your code review of my code, it made it better.

jfkisafk replied:
No worries, one additional thing I would suggest is to add another command to "Format Selected SQL". You can find that I am adding similar command in https://github.com/raycast/extensions/pull/14155. This saves an extra step (not copying to clipboard), if someone just wants to format a selected text.

tonngw replied:
I looked at the PR cited in the comment above, and it just so happened that I had the same idea as you, and I plan to add it in the next update. Thanks for your suggestions, I think getting from the clipboard and selecting text for formatting should be two very common operations.

jfkisafk commented on extensions/sql-format/src/index.tsx:
```diff
@@ -0,0 +1,179 @@
+
+// copy or auto paste output
+export async function copyFormattedSQL(result: string) {
+  if (autoPasteEnabled()) {
+    await Clipboard.paste(result);
+    await showHUD('✅ Pasted succesfully!');
+  } else {
+    await Clipboard.copy(result);
+    await showHUD('✅ Copied succesfully!');
+  }
```
(non-blocking) We can use toasts here in UserInitiated mode for better color coded prompts. See the changes in https://github.com/raycast/extensions/pull/14155 for more details.

tonngw replied:
It is more colorful  after I changed, but I want to know when LaunchType.Background to use it. Beside choose <run in background>, the action is not frequently.
<img width="750" alt="image" src="https://github.com/user-attachments/assets/682a27b5-3441-4106-8ca5-04f5d5c201c7">

jfkisafk replied:
You're right, in most cases these commands will not run in Background mode. But, to be fault-tolerant in case someone launches the command with [deeplink](https://developers.raycast.com/information/lifecycle/deeplinks) and `launchType=background`, the command will not fail (as Toasts are not available in Background mode).

tonngw replied:
ok, i got it, you explanation is very clear about it.

jfkisafk reviewed (approved):
LGTM. Just some minor comments for better UX.

tonngw commented on extensions/sql-format/CHANGELOG.md:
```diff
@@ -0,0 +1,5 @@
+# Changelog
+
+## [Added Format SQL] - 2024-08-23
```
I solved it.

jfkisafk replied:
Nice, please resolve all the comments you have solved already.

tonngw commented on extensions/sql-format/.prettierrc:
```diff
@@ -0,0 +1,4 @@
+{
+  "trailingComma": "all",
+  "singleQuote": true
+}
```
I solved it.

tonngw commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,96 @@
+    "@types/node": "20.8.10",
+    "@types/react": "18.3.3",
+    "eslint": "^8.57.0",
+    "prettier": "^2.8.8",
+    "ray": "^0.0.1",
```
I solved it.

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+    {
+      "data": [
+        {
+          "title": "Default: 2",
```
```suggestion
          "title": "Spaces: 2",
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+      "description": "Output sql tabWidth",
+      "name": "tabWidth",
+      "required": false,
+      "title": "tabWidth",
```
Title(s) should generally follow title case and defaults can be provided for preference, rather than specifying in name. Raycast will mark the default in this case.

```suggestion
      "title": "Tab Width",
      "default": "2",
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+      "description": "Output sql keywordCase",
+      "name": "keywordCase",
+      "required": false,
+      "title": "keywordCase",
```
```suggestion
      "title": "Keyword Case",
      "default": "preserve",
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+    {
+      "data": [
+        {
+          "title": "Default",
```
```suggestion
          "title": "Preserve Case",
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+          "value": "preserve"
+        },
+        {
+          "title": "Upper",
```
```suggestion
          "title": "Upper Case",
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+          "value": "upper"
+        },
+        {
+          "title": "Lower",
```
```suggestion
          "title": "Lower Case",
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,95 @@
+  },
+  "devDependencies": {
+    "@raycast/eslint-config": "^1.0.8",
+    "@types/node": "20.8.10",
+    "@types/react": "18.3.3",
```
You can run either `npm run i @types/{node@22,react@18} -D` or you can even remove these dependencies `npm uninstall @types/{node, react}`

jfkisafk commented on extensions/sql-format/src/index.tsx:
```diff
@@ -0,0 +1,173 @@
+      return;
+    }
+
+    const { tabWidth } = getPreferenceValues<Preferences>();
+    const { keywordCase } = getPreferenceValues<Preferences>();
```
```suggestion
    const { tabWidth, keywordCase } = getPreferenceValues<Preferences>();
```

jfkisafk reviewed (commented):
Sorry I missed package.json file changes and suggested a few refactors that you can apply straightaway from GitHub. But otherwise this command looks complete.

tonngw commented:
> Sorry I missed package.json file changes and suggested a few refactors that you can apply straightaway from GitHub. But otherwise this command looks complete.

It doesn't matter, I looked for newly your comments, refactored the project, and added a new command 「formatSelection」as you make suggesions. In addition, I updated the extension icon, what you provide is better than what I have used.  I feel that this extension is getting better and better, thank you for your efforts. I cited your github profile at README.md. :)

jfkisafk commented on extensions/sql-format/CHANGELOG.md:
```diff
@@ -0,0 +1,9 @@
+# Changelog
+
+## [Added Format Selected SQL command] - {PR_MERGE_DATE}
+
+- Added new command to format sql clause selected in the foremost editor.
+
+## [Added Format SQL] - {PR_MERGE_DATE}
+
+- Initial version, support format a SQL cluase string stored in clipboard and copies/pastes it back
```
```suggestion
## [Initial Version] - {PR_MERGE_DATE}

- Added a comment to format a SQL clause string stored in clipboard and copy/paste it back.
- Added new command to format sql clause selected in the foremost editor.
```

jfkisafk commented on extensions/sql-format/package.json:
```diff
@@ -0,0 +1,103 @@
+  "title": "Format SQL",
+  "description": "Format a SQL cluase use standard SQL format",
+  "icon": "extension-icon.png",
+  "author": "tonngw",
```
```suggestion
  "author": "tonngw",
  "contributors": ["stelo"],
```

tonngw replied:
I have added it, happy working with you.

jfkisafk commented on extensions/sql-format/metadata/sql-format.png:
Btw can you change this metadata img to use the one from this PR description (reflecting the new icon and command)?

tonngw replied:
yeah, this is indeed an inadvertent change, I give you a thumbs up. 👍

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥

tonngw commented:
@tonngw good job!

tgw2020 commented:
@tonngw nice

tgw2020 commented:
@tonngw nice!
